### PR TITLE
Adding browser outdated notification

### LINKED ISSUE #16283: Notify users that we are dropping support for their current browser
We would like to notify users when we are planning to drop support for the browser they are currently using, so that users have at least once release to learn about this change and how it will impact them.

### PR DESCRIPTION
**Pull request by vthomas13**
Adding a notification for users with outdated Chrome (<80) and Firefox (<78) browser versions that will lose support in the future. This notification should only show once every two days, while the browser remains outdated.

* Fixes #16283

## Manual Testing Steps
Download Chromium v78
[Windows](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/693951/), [macOS](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac/693954/), [Linux](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Linux_x64/693954/)
Install a MetaMask extension build from this branch.

## Before
image

## After
image

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes

## Pre-merge reviewer checklist

- [X] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [X] PR is linked to the appropriate GitHub issue
- [X] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

### CODE REVIEW

File: app/scripts/lib/createDupeReqFilterMiddleware.test.js
Comment (vthomas13) [inline app/scripts/lib/createDupeReqFilterMiddleware.test.js:28]: These changes were introduced from the yarn lint:fix command. I can take them out if necessary.

Comment (Gudahtt) [reply app/scripts/lib/createDupeReqFilterMiddleware.test.js:28]: Hmm. That shouldn't happen, but I did confirm this to be the case just because I was curious. Weird.

This problem does go away after rebasing/merging latest develop though.

Comment (vthomas13) [reply app/scripts/lib/createDupeReqFilterMiddleware.test.js:28]: okay, thanks for the insight @Gudahtt, I still need to get familiar with the pipeline

**Comment by vthomas13** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ui/helpers/constants/browser.js
Comment (jpuri) [inline ui/helpers/constants/browser.js:4]: May be we can use existing constant files and not add new just for browser constants.

File: ui/pages/home/home.component.js
Comment (jpuri) [inline ui/pages/home/home.component.js:252]: where is this state variable being used `outdatedBrowserWarningLastShown` ?

**Review by jpuri — DISMISSED** [review]
Looks good 👍

File: ui/selectors/selectors.js
Comment (jpuri) [inline ui/selectors/selectors.js:1056]: logic like above should be avoided in selector I think

File: jest.config.js
Comment (Gudahtt) [inline jest.config.js:42]: These all look like duplicate entries
```suggestion
```

File: ui/helpers/utils/util.test.js
Comment (Gudahtt) [inline ui/helpers/utils/util.test.js:198]: Could you add test cases for Edge and Opera as well?

I'm also curious how browser forks are treated here. Could you add a test for an unrecognized browser?

File: ui/pages/home/home.component.js
Comment (Gudahtt) [inline ui/pages/home/home.component.js:490]: It looks like you're passing a "Here" span into the outdated browser notification message, but there aren't any substitutions in that message. This syntax is for when we're embedding something in a message.

File: ui/store/actions.js
Comment (Gudahtt) [inline ui/store/actions.js:3278]: We should use the async `submitRequestToBackground` here instead. This `callBackgroundMethod` function should be marked as deprecated. We've been in the process of migrating away from using the callback-style background calls for years now.

Comment (Gudahtt) [reply ui/store/actions.js:3278]: Also, this is an async call so it should be an `async` function that waits for the call to complete. That makes it possible for the caller to wait, if they want to. Otherwise there's no way to tell when the request has completed.

File: app/scripts/controllers/app-state.test.js
Comment (vthomas13) [reply app/scripts/controllers/app-state.test.js]: Thanks, great point.

File: ui/pages/home/home.component.js
Comment (vthomas13) [reply ui/pages/home/home.component.js:490]: Thanks!

File: ui/store/actions.js
Comment (Gudahtt) [inline ui/store/actions.js:3280]: Nit: Could you use `try .. catch` here instead of `.catch`? The former will preserve an async stack trace across this line on Chromium-based browsers, leading to better error messages. Generally async/await should be preserved over .then/.catch in all cases for this reason.

File: ui/helpers/utils/util.test.js
Comment (vthomas13) [reply ui/helpers/utils/util.test.js:198]: Right now it looks like Opera is unhandled because bowser handles by name, but I've manually tested version 58 which was released Jan/23/2019 (around the same time as Chromium v78 and it seemed to work fine. I can add it to the list of outdated browser ranges as '<58'

Comment (Gudahtt) [reply ui/helpers/utils/util.test.js:198]: Could you clarify what you mean by "unhandled"? Does that mean it will _always_ say that the browser is outdated, or `_never_`? Not sure what the default case is.

Comment (vthomas13) [reply ui/helpers/utils/util.test.js:198]: The default case is the current functionality, where the user would not see anything different. So it will never say the browser is outdated if it's not in [this constant](https://github.com/MetaMask/metamask-extension/blob/938fd4ebaca5310c8a1b7b94214d11e383edc7c1/ui/helpers/constants/common.js#L27-L30)

Comment (Gudahtt) [reply ui/helpers/utils/util.test.js:198]: The Wikipedia page claims that Opera 65 was based on Chromium v78, so maybe we can set that as the minimum for Opera: https://en.wikipedia.org/wiki/History_of_the_Opera_web_browser

Comment (Gudahtt) [reply ui/helpers/utils/util.test.js:198]: Oh sorry, it's Chrome 80 we're targeting, which'd be Opera 67

>Opera 67 was released on February 25, 2020, based on Chromium 80.

File: ui/store/actions.js
Comment (Gudahtt) [inline ui/store/actions.js:3285]: This `displayWarning` method should be considered deprecated as well. That behavior shows the error directly to the error in a very confusing and nonsensical way.

Sorry when I suggested using `try..catch` I hadn't considered whether there was an error to catch in the first place. It's probably better to let this one bubble up.

```suggestion
    await submitRequestToBackground('setOutdatedBrowserWarningLastShown', [
      lastShown,
    ]);
    await forceUpdateMetamaskState(dispatch);
```

Comment (vthomas13) [reply ui/store/actions.js:3285]: Gotcha, I appreciate you explaining the rationale behind it too, thanks!

Comment (Gudahtt) [reply ui/store/actions.js]: Documented here: https://github.com/MetaMask/metamask-extension/pull/17465

**Review by Gudahtt — DISMISSED** [review]
LGTM, great work!

**Review by mcmire — DISMISSED** [review]
Looks good, just had a couple of suggestions, but you can take them or leave them.

File: ui/store/actions.js
Comment (Gudahtt) [reply ui/store/actions.js:3281]: Good point. It does update automatically, but it's async. We use this force update in cases where we need to guarantee that the state has been updated before we perform the next operation.

In this case there is no next operation though, so I don't think we need it

File: app/scripts/controllers/app-state.test.js
Comment (vthomas13) [reply app/scripts/controllers/app-state.test.js:3]: I think so too, I think the lint:fix command changed it to function notation instead of lambda notation

File: .eslintrc.js
Comment (Gudahtt) [inline .eslintrc.js:265]: It looks like this needs to be listed under the `excludedFiles` entry of the block above here. That's probably why the linter converted the test to use functions - it's a mocha rule, and that's still getting applied.

File: app/scripts/controllers/app-state.test.js
Comment (mcmire) [reply app/scripts/controllers/app-state.test.js:3]: Ah. Interesting. Okay!

File: .eslintrc.js
Comment (Gudahtt) [reply .eslintrc.js:265]: Probably should call out more explicitly that this list needs to be in-sync with the list in the previous block. I see there are other discrepancies as well

**Review by Gudahtt — DISMISSED** [review]
One pending suggestion, but we can handle that in a follow-up PR instead if you want

File: .eslintrc.js
Comment (vthomas13) [reply .eslintrc.js:266]: I could sync that up, but I think the formatting has to be changed to lambda notation in each test file manually, so it might be better left for a separate PR/changeset.

Comment (Gudahtt) [reply .eslintrc.js:266]: Oh yeah definitely, I agree

**Review by Gudahtt — COMMENTED** [review]
Looks great! I can re-approve when the conflicts are resolved

**Review by pedronfigueiredo — DISMISSED** [review]

**Review by Gudahtt — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
We reviewed this one ysterday live, and now that the change is made this looks good to me!

**Pull request merged**
